src/gui/sdlmain.cpp

```diff
@@ -1846,19 +1950,43 @@ static void SetupWindowResolution(const char *val)
 		SDL_Rect bounds;
 		SDL_GetDisplayBounds(sdl.display_number, &bounds);
 		if (w > 0 && h > 0 && w <= bounds.w && h <= bounds.h) {
+			sdl.desktop.want_resizable_window = false;
 			sdl.desktop.window.use_original_size = false;
 			sdl.desktop.window.width = w;
 			sdl.desktop.window.height = h;
 			return;
 		}
 	}
 
-	LOG_MSG("MAIN: 'windowresolution = %s' is not a valid setting, using "
-	        "'original' instead",
-	        pref.c_str());
+        LOG_MSG("MAIN: 'windowresolution = %s' is not a valid setting, using "
+                "'original' instead", pref.c_str());
+        sdl.desktop.want_resizable_window = false;
 	sdl.desktop.window.use_original_size = true;
 }
 
+static SDL_Rect CalculateViewport(int win_width, int win_height)
+{
+	const double prog_aspect_ratio = (sdl.draw.width * sdl.draw.scalex) /
+	                                 (sdl.draw.height * sdl.draw.scaley);
+	const double win_aspect_ratio = double(win_width) / double(win_height);
+
+	if (prog_aspect_ratio > win_aspect_ratio) {
+		// match window width
+		const int w = win_width;
+		const int h = win_width / prog_aspect_ratio;
+		const int x = 0;
```

Comment:
(Minor: given these are sizes and dimensions, logically they should never be negative and so we should use unsigned values; however SDL tends to deal in signed values.. for reasons I don't understand).

Reply:
They have at least 3 good reasons actually :) SDL 1 used signed 16-bit values for x an y and unsigned 16-bit values for width and height - on the cover it looks quite fine, as it closely matches maximum resolutions available in most OSes (e.g. max 16kx16k in Windows; for X11 I think it's 65kx65k but I doubt anything beyond 16k would actually work), but in practice:

1. OpenGL uses GLint for all those anyway (which is signed 32-bit integer)
2. Storing width and height in 16-bit type can easily cause overflows during various calculations (interestingly, DOSBox devs came to the same conclusion! sizes of various VGA resolutions are stored in Bitu even when all max resolutions are known at compile time and much smaller - see `struct VideoModeBlock` and big table of modes in `src/ints/int10_modes.cpp`).
3. In C/C++ arithmetic operation on signed and unsigned type with same size always gives unsigned type in the result. This can very easily happen when calculation uses both x,y and width/height.

Using simple `int` for all of these solves all 3 problems, with additional bonus of future-proofing the code (8k displays are the reality now, so we're closing in on maximum 16k OS limits).

Reply:
Thanks for the explanation and background; makes sense!